### PR TITLE
Fix android copy paste menu doens't show up in compose box

### PR DESCRIPTION
**Pull request by K4tsuki**
### Details

Fix copy paste context menu won't show up in android.
Change `TextInput` of compose box into uncontrolled `TextInput` that will improve performance of `TextInput`

### Fixed Issues

$ https://github.com/Expensify/App/issues/6876

### Tests

1. Type something in compose box in Android
2. Select/highlight part of text 
3. Verify that Copy / Paste menu context show up in Android
4. Set selection/cursor at certain part of text 
5. Input an emoji
6. Verify emoji placement, cursor position and text change is correct after input an emoji
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Type something in compose box in Android
2. Select/highlight part of text 
3. Verify that Copy / Paste menu context show up in Android
4. Set selection/cursor at certain part of text 
5. Input an emoji
6. Verify emoji placement, cursor position and text change is correct after input an emoji
- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [X] Desktop
- [X] iOS
- [x] Android

#### Web

https://user-images.githubusercontent.com/1313124/153714274-f5061bd8-44a9-410f-9515-79c71b74a8aa.mp4

#### Mobile Web

https://user-images.githubusercontent.com/1313124/153714681-c847d869-6f94-4b0e-9cec-1148f9abe89e.mp4

#### Desktop

https://user-images.githubusercontent.com/1313124/154658222-90880932-b71a-4e40-871a-7e0fd51d7ee8.mp4

#### iOS

https://user-images.githubusercontent.com/1313124/154636904-d2175507-edcb-4327-b853-9f7ba679cf16.mp4

#### Android

https://user-images.githubusercontent.com/1313124/153714288-f30483e7-9ca3-484c-94ea-1a2a8580509b.mp4

### CODE REVIEW

File: src/pages/home/report/ReportActionCompose.js [hunk 1]
```diff
@@ -310,18 +351,18 @@ class ReportActionCompose extends React.Component {
...
+                this.setState({isCommentEmpty: false});
+            }
             Report.setReportWithDraft(this.props.reportID.toString(), true);
         }
 
         // The draft has been deleted.
         if (newComment.length === 0) {
+            if (!this.state.isCommentEmpty) {
+                this.setState({isCommentEmpty: true});
+            }
```
Comment (parasharrajat) [inline src/pages/home/report/ReportActionCompose.js:365]: I didn't understand this part. Also, I would like to ask that your original solution is very different from your proposal. Why so?

File: src/pages/home/report/ReportActionCompose.js [hunk 2]
```diff
@@ -241,24 +245,61 @@ class ReportActionCompose extends React.Component {
...
     }
 
+    /**
+     * Set text input value and selection, then call updateComment with newText parameter
+     *
+     * @param {String} newText
+     * @param {Object} selection: {start: number, end: number}
+     */
+    setTextAndSelectionOfTextInput(newText, selection) {
+        if (Platform.OS === 'web') {
```
Comment (parasharrajat) [inline src/pages/home/report/ReportActionCompose.js:255]: We don't use this. [Check this](https://github.com/Expensify/App#platform-specific-file-extensions).

Comment (K4tsuki) [reply src/pages/home/report/ReportActionCompose.js:255]: Alright, I will try other way.

File: src/pages/home/report/ReportActionCompose.js (hunk 1 above)
Comment (K4tsuki) [reply src/pages/home/report/ReportActionCompose.js:365]: Because I realize `updateCommnet` is handler for `onChangeText` and put `setNativePropse(text: )` in there isn't doing anything because when user type a character setNativeProps is being called even though the textInput is already changed by user type (you don't need `setNativeProps` there). 

As for `setState`, each time user type a character `setState` is being called then `ReportActionCompose` is being rendered. I am guarding `setState` of `isCommentEmpty` so it is not being called unnecessarily.

https://stackoverflow.com/questions/24718709/reactjs-does-render-get-called-any-time-setstate-is-called

Here I am attaching performance improvement, the first compose box is build using `developement` build but  it is faster than second compose box that is current production release.

https://user-images.githubusercontent.com/1313124/153715911-6d68ef6f-c4e2-48e7-abba-72c54cc96aee.mp4

Comment (parasharrajat) [reply src/pages/home/report/ReportActionCompose.js:365]: Ok. I think we can refactor this into a small function so that it does not look confusing.
```js
setCommentEmpty(IsEmpty){
	  if(this.state.isCommentEmpty === IsEmpty}{
	  	return;
	  }
	  this.setState({isCommentEmpty: IsEmpty});
}
```

File: src/pages/home/report/ReportActionCompose.js [hunk 3]
```diff
@@ -241,24 +245,61 @@ class ReportActionCompose extends React.Component {
...
+            // react-native-web
+            // because setNativeProps in react-native-web is slow and doesn't have good support
+            this.textInput.value = newText;
+            this.textInput.setSelectionRange(selection.start, selection.end);
+        } else {
+            this.textInput.setNativeProps({text: newText, selection});
+
+            // Relasing selection handling to system
+            setTimeout(() => this.textInput.setNativeProps({selection: {start: undefined, end: undefined}}), 0);
+        }
```
Comment (parasharrajat) [inline src/pages/home/report/ReportActionCompose.js:265]: Let's discuss this first. I have a couple of questions.
1. why `because setNativeProps in react-native-web is slow and doesn't have good support`? Any reasons to support your comment.
2. Use of `setTimeout`. Why is this needed?

File: src/pages/home/report/ReportActionCompose.js (hunk 1 above)
Comment (K4tsuki) [reply src/pages/home/report/ReportActionCompose.js:365]: All right.

File: src/pages/home/report/ReportActionCompose.js (hunk 3 above)
Comment (K4tsuki) [reply src/pages/home/report/ReportActionCompose.js:265]: It is `setTimeout` 0, you can read more about it here:
https://stackoverflow.com/questions/779379/why-is-settimeoutfn-0-sometimes-useful
For this code it is to ensure that the previous `setNativeProps` is done then do second `setNativeProps`. 

As for `setNativeProps` for react-native-web, please wait.

Comment (parasharrajat) [reply src/pages/home/report/ReportActionCompose.js:265]: And this is very confusing that you are first settings the selection and then releasing it.

Comment (K4tsuki) [reply src/pages/home/report/ReportActionCompose.js:265]: If we don't do that, the selection will stay or come back at that position.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
There is an issue with IOS. Where after adding the emoji the cursor is moved to the start of the text. 

Please test it on IOS and Destop and add videos for those.

**Comment by K4tsuki** [comment]
@parasharrajat  Add dekstop and ios videos, and add new commit.

**Comment by K4tsuki** [comment]
I am able to find another solution for IOS without second `setNativeProps` and without `setTimeout` / `Promise resolve`.

**Comment by K4tsuki** [comment]
@stitesExpensify for update, the `setTimeout 0` could be replaced with `Promise/resolve`.

**Comment by parasharrajat** [comment]
I have raised some concerns on the slack. I will revisit this in some days.

File: src/libs/TextInputUtils/index.ios.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/**
+* Set value and selection of uncontrolled TextInput element
+*
+* @param {Object} textInputElement
+* @param {String} text
+* @param {Object} selection: {start: number, end: number}
+*/
+function setTextAndSelection(textInputElement, text, selection) {
```
Comment (parasharrajat) [inline src/libs/TextInputUtils/index.ios.js:8]: @stitesExpensify This function is implemented differently on Android|IOS|Web. What do you think about this imlementation?

Comment (stitesExpensify) [reply src/libs/TextInputUtils/index.ios.js:8]: I think I'm missing some context. Why does this need to be implemented in 3 different ways @K4tsuki ?

Comment (K4tsuki) [reply src/libs/TextInputUtils/index.ios.js:8]: @stitesExpensify for the web, setNativeProps for `selection` doesn't work, because to set the selection of text input in web the property name is `selectionStart` and `selectionEnd`. So it has different name than native `selection` property.

For IOS, at first I am using android code, but  there is a bug, like what @parasharrajat mention above.
> There is an issue with IOS. Where after adding the emoji the cursor is moved to the start of the text.

The native behavior of setting selection using `setNativeProps` is different in IOS and android, so I am just adjusting the code based on native behavior of android and IOS.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Issue (Web): NumberLines is not updated which means the necessary events are not fired when we update the value via `setTextAndSelection`.

Steps: 
1. Write a multiline comment for about 6 lines.
2. Select the text on lines 3-5. 
3. add an emoji. 
4. Notice that emoji is placed fine but there are extra lines on the text input. 
5. But when we start typing the lines are recalc. 

Lines should update when we update the value on input.

**Review by parasharrajat — COMMENTED** [review]
Ok. The solution started to become messy and hacky. I will try to do some digging first. But We are just trying to patch in code to fix issues that are going to introduce more issues. 

I would not suggest going too deep with platform-specific code.  
As events are not firing, it means that we are doing it all wrong. Although we can put in all the vanilla js in react we should avoid it. Instead, do things react way. 

Let me know if you have other ideas as well. I would hear them out first.

**Comment by K4tsuki** [comment]
@parasharrajat What is the ideal solution? Should we don;t use platform specific code at all for this issue?

**Comment by K4tsuki** [comment]
@stitesExpensify What is the ideal solution here? Could I use my code inside `TextInputFocusable`, because the component have platform specific code, so I am not making new platform specific code, but  I will use the existing component.

Also, Are `setTimeout 0` and platform specific code strictly prohibited here? 

Also there is something in:

https://github.com/Expensify/App/blob/bc4f429849f87da71962638c204351a946cee474/src/components/TextInputFocusable/index.js#L339-L343

As you could see, there is a `setState` callback for `setState`. Is this allowable while my `setTimeout 0` prohibited?

**Comment by parasharrajat** [comment]
I will give it a shot and see if something else can be done here. It is fine to use platform-specific code but this should only be used when there are no other choices. `setNativeProps` is necessary to use. If you remember the whole problem is that the menu does not show due to controlled prop. 

Apart from it, the problem is not that we are using platform-specific code instead of the way code is used.

**Comment by stitesExpensify** [comment]
Closing this since K4tsuki is no longer a contributor
